src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -80,30 +95,36 @@ SchedulingResult GcsResourceScheduler::Schedule(
     return std::make_pair(SchedulingResultStatus::INFEASIBLE, std::vector<NodeID>());
   }
 
+  // If scheduling type is strict pack, we do not need to sort resources of each placement
+  // group since they must exist on the same node
+  if (scheduling_type == SchedulingType::STRICT_PACK) {
+    return StrictPackSchedule(required_resources_list, candidate_nodes);
+  }
+
   // First schedule scarce resources (such as GPU) and large capacity resources to improve
   // the scheduling success rate.
-  const auto &to_schedule_resources = SortRequiredResources(required_resources_list);
+  const auto &sorted_index = SortRequiredResources(required_resources_list);
+
+  std::vector<ResourceSet> sorted_resources(required_resources_list);
+  for (size_t i = 0; i < sorted_index.size(); i++) {
+    sorted_resources[i] = required_resources_list[sorted_index[i]];
+  }
 
   // Score and rank nodes.
-  SchedulingResult result;
   switch (scheduling_type) {
+  case PACK:
+    return SortSchedulingResult(PackSchedule(sorted_resources, candidate_nodes),
+                                sorted_index);
   case SPREAD:
-    result = SpreadSchedule(to_schedule_resources, candidate_nodes);
-    break;
+    return SortSchedulingResult(SpreadSchedule(sorted_resources, candidate_nodes),
+                                sorted_index);
   case STRICT_SPREAD:
-    result = StrictSpreadSchedule(to_schedule_resources, candidate_nodes);
-    break;
-  case PACK:
-    result = PackSchedule(to_schedule_resources, candidate_nodes);
-    break;
-  case STRICT_PACK:
-    result = StrictPackSchedule(to_schedule_resources, candidate_nodes);
-    break;
+    return SortSchedulingResult(StrictSpreadSchedule(sorted_resources, candidate_nodes),
+                                sorted_index);
   default:
     RAY_LOG(FATAL) << "Unsupported scheduling type: " << scheduling_type;
-    break;
+    return SchedulingResult();
```

Comment:
We don't actually need this since we cannot reach here (FATA will crash the process).

Maybe instead of RAY_LOG(FATAL), you can use UNREACHABLE https://github.com/ray-project/ray/blob/2302b4eea87c480699797e4bf770d46d0b758379/src/ray/raylet/worker_pool.cc#L1506